roles/pool/src/lib/mod.rs

```diff
@@ -3,39 +3,49 @@ pub mod mining_pool;
 pub mod status;
 pub mod template_receiver;
 
+use std::sync::Arc;
+
 use async_channel::{bounded, unbounded};
 
+use error::PoolError;
 use mining_pool::{get_coinbase_output, Configuration, Pool};
+use roles_logic_sv2::utils::Mutex;
 use template_receiver::TemplateRx;
 use tracing::{error, info, warn};
 
 use tokio::select;
 
+#[derive(Debug, Clone, PartialEq)]
+pub enum RoleState {
```

Comment:
interesting choice on `enum RoleState` :thinking: :thought_balloon: 

so I assume there will be more commits replicating this pattern on the other roles?

Reply:
Didn't we discussed to not have code changes specific to integration test in the production code for roles?

Reply:
@plebhash yea the motivation is to have this shared between the different roles, anyhow for this PR Ill just change it to `PoolState` and change it(and move it out to a shared crate) when needed

Reply:
> Didn't we discussed to not have code changes specific to integration test in the production code for roles?

I get where you are coming from and I'll try to elaborate more what I stated in https://github.com/stratum-mining/stratum/issues/1148#issuecomment-2315595364, but by the way you constructed your question I feel I should first start with some clarifications:

`roles` is not production code... in fact, quite the opposite: it should be seen as a mere PoC that we use to showcase the community that SV2 works (e.g.: is able to mine blocks with JD).

what we should see as production code is `protocols`, which are the building blocks that we expect industry players to use while building their own solutions.

---

when it comes to changing `roles` for Integration Tests, as I stated, we should aim to be as conservative as possible... if we try to change their internal architecture, the scope of the effort can grow exponentially and put us in a situation where this effort is no longer worth it

for example, if we were aiming to change the architecture around `Status` and all the related channels, then I would feel this was a pandora box not really worth opening

but the solution @jbesraa came up with here is relatively straightforward (just add an extra `enum`) and I haven't detected any big refactoring implications as a consequence, so I'm not going to oppose it

---

but again, overall I get where you're coming from @Shourya742 and we will probably have similar discussions again when we get to #1122

Reply:
Oh!!! I thought the roles are production code. Coming up with state machine was never an issue, but what I remembered the issue was that it gonna lead to deep refactor in roles. I still haven't reviewed the PR, but if it doesn't changes too much stuff then we are good go.